Sources/PostgresNIO/New/PSQLRowSequence.swift

```diff
@@ -0,0 +1,607 @@
+        typealias Element = PSQLRow
+        
+        let _internal: _Internal
+        
+        init(consumer: AsyncStreamConsumer) {
+            self._internal = _Internal(consumer: consumer)
+        }
+        
+        mutating func next() async throws -> PSQLRow? {
+            try await self._internal.next()
+        }
+        
+        final class _Internal {
+            let consumer: AsyncStreamConsumer
+            
+            init(consumer: AsyncStreamConsumer) {
+                self.consumer = consumer
+            }
+
+            deinit {
+                self.consumer.iteratorDeinitialized()
+            }
+            
+            func next() async throws -> PSQLRow? {
+                try await self.consumer.next()
+            }
+        }
+    }
+}
+
+final class AsyncStreamConsumer {
+    let lock = Lock()
+    
+    let lookupTable: [String: Int]
+    let columns: [RowDescription.Column]
+    let jsonDecoder: PSQLJSONDecoder
+    private var state: StateMachine
+    
+    init(
+        lookupTable: [String: Int],
+        columns: [RowDescription.Column],
+        jsonDecoder: PSQLJSONDecoder
+    ) {
+        self.state = StateMachine()
+        
+        self.lookupTable = lookupTable
+        self.columns = columns
+        self.jsonDecoder = jsonDecoder
+    }
+    
+    func startCompleted(_ buffer: CircularBuffer<DataRow>, commandTag: String) {
+        self.lock.withLock {
+            self.state.finished(buffer, commandTag: commandTag)
+        }
+    }
+    
+    func startStreaming(_ buffer: CircularBuffer<DataRow>, upstream: PSQLRowStream) {
+        self.lock.withLock {
+            self.state.buffered(buffer, upstream: upstream)
+        }
+    }
+    
+    func startFailed(_ error: Error) {
+        self.lock.withLock {
+            self.state.failed(error)
+        }
+    }
+    
+    func receive(_ newRows: [DataRow]) {
+        let receiveAction = self.lock.withLock {
+            self.state.receive(newRows)
+        }
+        
+        switch receiveAction {
+        case .succeed(let continuation, let data, signalDemandTo: let source):
+            let row = PSQLRow(
+                data: data,
+                lookupTable: self.lookupTable,
+                columns: self.columns,
+                jsonDecoder: self.jsonDecoder
+            )
+            continuation.resume(returning: row)
+            source?.demand()
+            
+        case .none:
+            break
+        }
+    }
+    
+    func receive(completion result: Result<String, Error>) {
+        let completionAction = self.lock.withLock {
+            self.state.receive(completion: result)
+        }
+        
+        switch completionAction {
+        case .succeed(let continuation):
+            continuation.resume(returning: nil)
+            
+        case .fail(let continuation, let error):
+            continuation.resume(throwing: error)
+            
+        case .none:
+            break
+        }
```

Comment:
Is there really no less mind-numbingly tedious way to forward the result to the completion than this? 😕 I'm not even snarking (for once 😅 - Having to scroll through this kind of boilerplate when dealing with bugs and such surprisingly demoralizing in Swift (at lest, for me).

Reply:
I'm not sure I get what you would like to see here?

Reply:
I like the more verbose way of doing things here... My brain hurts when we use short syntax, and I want to separate concerns explicitly. Since I have written most of this and am probably the one to fix any issues, I'd prefer if we keep it as is.